core/src/storage/lazy.rs

```diff
@@ -0,0 +1,344 @@
+
+/// A lazy storage entity.
+///
+/// This loads its value from storage upon first use.
+///
+/// # Note
+///
+/// Use this if the storage field doesn't need to be loaded in some or most cases.
+#[derive(Debug)]
+pub struct Lazy<T> {
+    kind: UnsafeCell<LazyKind<T>>,
+}
+
+impl<T> StorageSize for Lazy<T>
+where
+    T: StorageSize,
+{
+    const SIZE: u64 = <T as StorageSize>::SIZE;
+}
+
+impl<T> Pull for Lazy<T>
+where
+    T: StorageSize + scale::Decode,
+{
+    fn pull(key_ptr: &mut KeyPtr) -> Self {
+        Self::lazy(key_ptr.next_for::<T>())
+    }
+}
+
+impl<T> Push for Lazy<T>
+where
+    T: Push,
+{
+    fn push(&self, key_ptr: &mut KeyPtr) {
+        // We skip pushing to contract storage if we are still in unloaded form.
+        if let LazyKind::Occupied(occupied) = self.kind() {
+            occupied.value.push(key_ptr)
+        }
+    }
+}
+
+impl<T> Lazy<T> {
+    /// Creates an eagerly populated lazy storage value.
+    #[must_use]
+    pub fn new(value: T) -> Self {
+        Self {
+            kind: UnsafeCell::new(LazyKind::Occupied(OccupiedLazy::new(value))),
+        }
+    }
+
+    /// Creates a true lazy storage value for the given key.
+    #[must_use]
+    pub fn lazy(key: Key) -> Self {
+        Self {
+            kind: UnsafeCell::new(LazyKind::Vacant(VacantLazy::new(key))),
+        }
+    }
+
+    /// Returns a shared reference to the inner lazy kind.
+    #[must_use]
+    fn kind(&self) -> &LazyKind<T> {
+        // SAFETY: We just return a shared reference while the method receiver
+        //         is a shared reference (&self) itself. So we respect normal
+        //         Rust rules.
+        unsafe { &*self.kind.get() }
+    }
+
+    /// Returns an exclusive reference to the inner lazy kind.
+    #[must_use]
+    fn kind_mut(&mut self) -> &mut LazyKind<T> {
+        // SAFETY: We just return an exclusive reference while the method receiver
+        //         is an exclusive reference (&mut self) itself. So we respect normal
+        //         Rust rules.
+        unsafe { &mut *self.kind.get() }
+    }
+
+    /// Performs the given closure on the mutable lazy kind.
+    ///
+    /// # Note
+    ///
+    /// Actions on the mutable lazy kind are performed within the closure
+    /// to not leak exclusive references to it to the outside. This is important
+    /// since the `for_kind` method itself operates only on `&self`.
+    fn for_kind<F, R>(&self, f: F) -> R
+    where
+        F: for<'a> FnOnce(&'a mut LazyKind<T>) -> R,
+    {
+        // SAFETY: We operate on an exclusive reference on `LazyKind` within the
+        //         given closure while our method receiver is only a shared reference.
+        //         However, due to encapsulation of the exclusive reference within
+        //         the given closure we cannot leak the exclusive reference outside
+        //         of the closure. So the below action is safe in this regard.
+        f(unsafe { &mut *self.kind.get() })
```

Comment:
Is this an adequate proof though? I think you can still get a mutable reference, for example:

```rust
lazy.for_kind(|a| lazy.for_kind(|b| {}));
```

Reply:
Yep, completely broken ... removed the internal utility function.